### PR TITLE
Fix GRPO chunked log-softmax on multi-GPU, and bound its peak memory

### PR DESCRIPTION
**Pull request by danielhanchen**
## Problem

`chunked_hidden_states_selective_log_softmax` does

```python
chunk_logits = chunk_hidden_states.to(lm_head.dtype) @ lm_head.t()
```

which assumes the hidden states and the output head are on the same device. When a model is dispatched across several GPUs they are not. accelerate places the tail of the model on the last device it fills, so `lm_head` can sit on `cuda:1` while the hidden states arrive from earlier layers on `cuda:0`, and the step dies with

```
Unhandled FakeTensor Device Propagation for aten.mm.default,
found two different devices cuda:0, cuda:1
```

Any GRPO run on a model too large for one card hits this. It reproduces on current `main`.

## Fix

Co-locate both operands on the head's device before the matmul, and send the per-token logprobs back to the caller's device so the concatenation and every downstream consumer see the device they started on.

On a single device every added `.to()` is a no-op and the result is bit-identical.

## Second change: bounding the peak

Each chunk materialises `rows x vocab` logits and then a float32 copy of them, all on the device holding the head. The chunk count is fixed at 4, so that peak grows with the batch rather than staying bounded. On a large vocabulary this is the difference between fitting and not.

`UNSLOTH_GRPO_MAX_ROWS_PER_CHUNK` caps the rows per chunk instead. It is pure loop splitting: more, smaller chunks, identical concatenated result. Measured on a 202k vocabulary at 3328 rows, peak on the head's device went from 994 MiB to 493 MiB.

Unset is the default and reproduces the previous chunk boundaries exactly, so this cannot change any existing run.

## Verification

Tests are in `tests/test_chunked_log_softmax_device.py`.

Against stock `main`, the two multi-GPU tests fail with the exact device mismatch and the single-device tests pass:

```
FAILED test_head_on_another_device_matches_single_device
FAILED test_index_on_caller_device_is_accepted
```

With this change, 7 passed.

The multi-GPU test asserts the split-device result is `torch.equal` to the single-device result, not merely close, and that the output comes back on the caller's device. The cap tests assert the unset default is `torch.equal` to before.

End to end, a 30B 4-bit GRPO run on 2 x 16 GB that previously died at the first optimizer step now completes, with the peak at 14.32 GiB of 14.56 GiB.

### CODE REVIEW

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -94,13 +94,36 @@ def chunked_hidden_states_selective_log_softmax(
...
     flat_index = index.reshape(-1)
 
+    # Each chunk materialises rows x vocab logits and then a float32 copy of
+    # them, all on the device holding the output head. With a large vocabulary
+    # and a fixed chunk count that grows with the batch, so the peak scales with
+    # the batch rather than staying bounded. Setting
+    # UNSLOTH_GRPO_MAX_ROWS_PER_CHUNK caps the rows per chunk instead, which is
+    # pure loop splitting: more, smaller chunks, same concatenated result.
+    # Unset (the default) keeps the previous chunk boundaries exactly.
+    max_rows = int(os.environ.get("UNSLOTH_GRPO_MAX_ROWS_PER_CHUNK", "0"))
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Correct, and confirmed on the declared floor. Reproduced by pasting the function source into the minimal namespace the generated trainer module gives it (torch + os) with the real fullgraph decorator:

- torch 2.4.0, variable unset: `Unsupported: constant fold exception: KeyError(b'UNSLOTH_GRPO_MAX_ROWS_PER_CHUNK')`
- torch 2.4.0, variable set: `Unsupported: const method call bytes.decode`, out of `os._Environ.__getitem__`

`_fall_back_to_eager_on_recompile_limit` only catches recompile-limit and disabled-hook breaks, so it propagated, exactly as you say.

Fixed in 9dfd268: the read moved into the signature default, evaluated once when the def runs, so nothing environment-related is traced. It also became a plain int argument that Dynamo guards on, which fixed a second problem the same experiment turned up: 2.7.1 and 2.9.1 constant-folded the read without a guard, so a value changed after the first trace was silently ignored, while 2.5.1 recompiled. A non-numeric value is now ignored rather than raised on.

Per-version, on the same pasted source: 2.4.0 fails before and passes after; 2.5.1, 2.6.0, 2.7.1, 2.8.0, 2.9.1 are bit-identical before and after.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_chunked_log_softmax_device.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    same = chunked_hidden_states_selective_log_softmax(hidden_states, lm_head, index, **KWARGS)
+
+    # The head is the tensor that moves: accelerate puts the tail of the model
+    # on the last device it fills, while the hidden states arrive from earlier
+    # layers on an earlier device.
+    split = chunked_hidden_states_selective_log_softmax(
+        hidden_states, lm_head.to("cuda:1"), index, **KWARGS,
+    )
+    assert split.device == hidden_states.device, "result must come back on the caller's device"
+    assert torch.equal(same, split)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_chunked_log_softmax_device.py:90]: Not taking this one. `torch.equal` is the assertion, not an accident: the whole claim of the device change is that co-locating on the head device is bit-identical to the single-device path, and `allclose` would let a real regression through. A dtype slip in the added `.to()`, for instance, lands well inside any tolerance loose enough to absorb a cross-architecture kernel difference.

Measured on this box, two GPUs, reference on cuda:0 against the head moved to cuda:1:

| dtype | torch.equal | max abs diff |
| --- | --- | --- |
| float32 | True | 0.0 |
| bfloat16 | True | 0.0 |
| float16 | True | 0.0 |

The scenario also needs two different GPU architectures in one accelerate dispatch, which is not a configuration this path is used in, and CI never reaches the test at all: both multi-GPU cases are skipped under `torch.cuda.device_count() < 2`. If a genuinely heterogeneous host ever does trip it, that is worth seeing rather than tolerating, because on such a host the single-device and split-device runs really would disagree and the caller should know.

**Comment by danielhanchen** [comment]
@codex review
